_posts/2020-03-25-#18238.md

```diff
@@ -10,7 +10,64 @@ status: upcoming
 commit:
 ---
 
-_Notes and questions to follow._
+## Notes
+- Transaction relay is a three step process: `INV` -> `GETDATA` -> `TX`. `INV`
+  messages are used to announce new transactions. `GETDATA` messages are used
+  by nodes to request specific transactions. And the transactions are delivered
+  with `TX` messages. If a node can no longer deliver the transaction, it will
+  respond with `NOTFOUND` instead of the `TX`.
+
+  You can learn more about these messages [here](https://en.bitcoin.it/wiki/Protocol_documentation#inv).
+
+- A node will not have more than one `GETDATA` in flight for a particular
+  transaction. The [TxDownloadState struct](https://github.com/bitcoin/bitcoin/blob/master/src/net_processing.cpp#L340)
+  is used to manage the state of transactions for each peer.
+
+- If a `GETDATA` request is out, a node will wait upto 2 minutes before timing
+  out and attempting to request the transaction from another peer.
+
+- Currently, if a peer responds with a `NOTFOUND`, the node doesn't use that
+  information to speedup a request to another peer. Instead, it continues to
+  wait until the 2 minute period is over.
+
+  This PR introduces logic to speedup a request to another peer when possible.
+
+- [PR 15505](https://github.com/bitcoin/bitcoin/15505) was a previous attempt
+  at implementing a solution, and was covered in a previous [PR Review Club
+  #15505](https://bitcoincore.reviews/15505.html). Reading the notes / logs /
+  PR can help shine light on some of the nuances that make this feature
+  tricky to implement.
+
+- Having this retry speedup is [mentioned](https://github.com/bitcoin/bitcoin/pull/17303#issuecomment-547589047) as a pre-req for [PR 17303](https://github.com/bitcoin/bitcoin/pull/17303).
+  That PR is trying to remove `mapRelay`, a data structure that keeps track of
+  recently announced transactions. Doing so would increase the likelihood of
+  receiving a `NOTFOUND` message. For example, nodes with small mempools could announce a
+  transaction, but then evict it before receiving the `GETDATA`. Without logic
+  for the receiving node to quickly retry, the node with a small mempool could be wasting
+  bandwidth, aka accidentally DoS-ing its peer.
+
+## Questions
+
+1. Did you review the PRs? [Concept ACK, approach ACK, tested ACK, or
+NACK?](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review)
+(Don't forget to put your PR review on GitHub.)
+
+2. What are some scenarios that lead to `NOTFOUND` messages?
+
+3. How does this PR implement the logic to speedup retries?
```

Comment:
as above, s/speedup/speed up/ (it's weird, but speed up is the verb, speedup is the noun)

Reply:
TIL